### PR TITLE
Autograd Function Fallback bug fix - moe support

### PR DESCRIPTION
**Pull request by pengwa**
**Description**: Autograd Function Fallback bug fix - moe support

- Support forward inputs orders like "Non_tensor/Tensor/Non_tensor". Correspondingly, support "None/Tensor_Grad/None" for backward outputs.
- Report RuntimeError when PythonOp detected but _enable_custom_autograd_function is NOT enabled.
- Simplify the attributed used by PythonOpGrad. Renaming some attribute to reflect whether it is used for tensor inputs or all (tensor + non-tensor) inputs.

Attached the PythonOP/PythonOpGrad schemas (and how those attributes used) after the change:

image

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

File: orttraining/orttraining/python/training/ortmodule/__init__.py [hunk 1]
```diff
@@ -41,7 +41,7 @@
 elif os.path.exists(os.path.join(TORCH_CPP_BUILD_DIR,'lock')):
     print("WARNING: ORTModule detected PyTorch's CPP extension lock file during initialization, "
           "which can cause the script to stop responding. "
-          f"Delete {os.path.join(TORCH_CPP_BUILD_DIR,'lock')} if a hang occurs.")
+          f"Delete {os.path.join(TORCH_CPP_BUILD_DIR,'lock')} if the script stop responding and not exiting.")
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule/__init__.py:44]: This us unrelated to this PR and it is being completely reworked at https://github.com/microsoft/onnxruntime/pull/8027/files. Try not changing this code to prevent conflicts between both PRs

Comment (pengwa) [reply orttraining/orttraining/python/training/ortmodule/__init__.py:44]: Sure, thanks for the heads up.

Comment (pengwa) [reply orttraining/orttraining/python/training/ortmodule/__init__.py]: Fixed, please check.

File: orttraining/orttraining/core/graph/training_op_defs.cc [hunk 2]
```diff
@@ -3123,38 +3117,12 @@ Return true if all elements are true and false otherwise.
...
-        // "unused" inputs are just control inputs and they are not used actual computation.
-        // Other variables are used to invoke autograd.Function.backward(ctx, grad_input1, ..., grad_input_N).
-        // The "input_count" here means 1 + N.
-        const auto input_count = input_tensor_requires_grads->ints().size();
-        ORT_ENFORCE(input_tensor_types_proto->ints_size() == input_count - 1,
-                    "PythonOp's input list should have one more element than \"input_tensor_types\" attribute.");
-        // The first input is a pointer which points to
-        // a Python object created by torch.autograd.Function.apply.
-        // For details, see how we interpret it in PythonOpGrad implementation.
-        for (auto i = 1; i < input_count; ++i) {
```
Comment (wschin) [inline orttraining/orttraining/core/graph/training_op_defs.cc:3143]: This doesn't look robust. I prefer to keep as many type checks as possible.

File: onnxruntime/core/language_interop_ops/torch/torch_proxy.cc [hunk 3]
```diff
@@ -171,6 +171,12 @@ void InvokeRunner(
   // first element.
   for (; i < static_cast<size_t>(PyTuple_Size(result_ptr.get())); ++i) {
     PyObject* dl_tensor_pointer = PyTuple_GetItem(result_ptr.get(), i);
+    if (dl_tensor_pointer == Py_None) {
```
Comment (wschin) [inline onnxruntime/core/language_interop_ops/torch/torch_proxy.cc:174]: We previously filter out all `None`s from Python side. Does this change means we don't filter out `None`s anymore?

Comment (pengwa) [reply onnxruntime/core/language_interop_ops/torch/torch_proxy.cc:174]: Yes, we don't filter them out, because None can be in any places of the output lists. We need map them correctly into PythonOp kernels's outputs.

File: orttraining/orttraining/core/graph/training_op_defs.cc (hunk 2 above)
Comment (pengwa) [reply orttraining/orttraining/core/graph/training_op_defs.cc:3143]: I think it is fine, because this is the shape inference of PythonOpGrad, all those attributes for example input_tensor_requires_grads comes from PythonOp's attribute. When we create PythonOpGrad, we already do the check. Does that make sense?

File: onnxruntime/core/language_interop_ops/torch/torch_proxy.cc [hunk 4]
```diff
@@ -171,6 +171,12 @@ void InvokeRunner(
...
+      OrtValue empty_ort_value;
+      returned_ortvalues.push_back(empty_ort_value);
+      continue;
+    }
+
     ORT_ENFORCE(Py_REFCNT(dl_tensor_pointer) == 1, "Ref count of dl_tensor_pointer should be 1.");
```
Comment (tlh20) [inline onnxruntime/core/language_interop_ops/torch/torch_proxy.cc:180]: Can the user's code cause a > 1 ref count here?

File: orttraining/orttraining/core/graph/gradient_builder.cc [hunk 5]
```diff
@@ -1805,9 +1772,24 @@ IMPLEMENT_GRADIENT_BUILDER(GetPythonOpGradient) {
     input_args.push_back(O(i));
   }
 
+  // src_attrs["input_requires_grads"] stores all inputs's requires_grad attributes,
+  // including both tensor inputs and non-tensor inputs (e.g. constants), here we filter out
+  // those non-tensor inputs when constructing PythonOpGrad's outputs.
```
Comment (tlh20) [inline orttraining/orttraining/core/graph/gradient_builder.cc:1777]: input_requires_grads is generated during export, right?  If so, do we have a mechanism to detect if the requirements change during training -- say if, for some reason, the graph is initially exported with more/fewer tensors requiring gradients than?  [ Not a use case I have seen, but looking to catch this kind of issue and report and error if it does occur ]

File: orttraining/orttraining/core/graph/training_op_defs.cc [hunk 6]
```diff
@@ -2887,9 +2887,14 @@ Return true if all elements are true and false otherwise.
...
       .Attr(
-          "call_convention",
-          "call_convention[i]==c means a non-tensor argument. call_convention[i]==d means a tensor.",
+          "input_convention",
+          "input_convention[i]==c means a non-tensor argument. input_convention[i]==d means a tensor.",
           AttributeProto::STRING)
+      .Attr(
+          "input_requires_grads",
+          "Flags to indicate whether the torch.autograd.apply's inputs (including both tensor"
+          " and non-tensor inputs) has gradient",
```
Comment (tlh20) [inline orttraining/orttraining/core/graph/training_op_defs.cc:2896]:
```suggestion
          "Flags to indicate whether the torch.autograd.apply's inputs require gradients (including flags for both tensor"
          " and non-tensor inputs)",
```

File: orttraining/orttraining/core/graph/training_op_defs.cc [hunk 7]
```diff
@@ -3063,8 +3064,10 @@ Return true if all elements are true and false otherwise.
       .Input(
           1,
           "inputs",
-          "Inputs of autograd.Function.backward. There are 2*N inputs: \
-          N gradient inputs + N forward run activations of PythonOp (not including the context).",
+          "There are 2*N inputs: "
+          "  N gradient inputs(as inputs of autograd.Function.backward) + "
```
Comment (tlh20) [inline orttraining/orttraining/core/graph/training_op_defs.cc:3068]:
```suggestion
          "  N gradient inputs (as inputs of autograd.Function.backward) + "
```

File: orttraining/orttraining/core/graph/training_op_defs.cc [hunk 8]
```diff
@@ -3086,34 +3089,25 @@ Return true if all elements are true and false otherwise.
...
       .Attr(
           "output_tensor_types",
-          "Output types of autograd.Function.backward.",
+          "Output types of autograd.Function.backward outputs(including only tensor outputs).",
           AttributeProto::INTS,
           false)
       .Attr(
           "output_tensor_ranks",
-          "Output ranks of autograd.Function.backward.",
+          "Output ranks of autograd.Function.backward outputs(including only tensor outputs).",
```
Comment (tlh20) [inline orttraining/orttraining/core/graph/training_op_defs.cc:3099]:
```suggestion
          "Output ranks of autograd.Function.backward outputs (including only tensor outputs).",
```

File: orttraining/orttraining/core/graph/training_op_defs.cc [hunk 9]
```diff
@@ -3086,34 +3089,25 @@ Return true if all elements are true and false otherwise.
...
           AttributeProto::INTS,
           false)
       .Attr(
           "output_tensor_requires_grads",
-          "Flags to indicate which outputs has gradient",
+          "Flags to indicate which outputs(including only tensor outputs) has gradient.",
```
Comment (tlh20) [inline orttraining/orttraining/core/graph/training_op_defs.cc:3104]:
```suggestion
          "Flags to indicate which outputs have gradients (including only tensor outputs).",
```

File: orttraining/orttraining/core/graph/training_op_defs.cc [hunk 10]
```diff
@@ -3123,38 +3117,12 @@ Return true if all elements are true and false otherwise.
...
-
         // Load expected output types.
         const auto output_tensor_types_proto = ctx.getAttribute("output_tensor_types");
         ORT_ENFORCE(static_cast<size_t>(output_tensor_types_proto->ints_size()) == ctx.getNumOutputs(),
                     "PythonOpGrad's output list and \"output_tensor_types\" attribute should have the same length.");
         // This is a required field.
-        ORT_ENFORCE(output_tensor_types_proto, "PythonOp's must have \"output_tensor_types\" attribute.");
+        ORT_ENFORCE(output_tensor_types_proto, "PythonOpGrad's must have \"output_tensor_types\" attribute.");
         // Set inferred output types.
         static size_t rank_count = 0;
```
Comment (tlh20) [inline orttraining/orttraining/core/graph/training_op_defs.cc:3127]: Is "static" correct here (and above, for PythonOp)?

File: orttraining/orttraining/training_ops/cpu/torch/torch_custom_function_kernel_base.cc [hunk 11]
```diff
@@ -30,11 +30,11 @@ void PythonOpBase::Init(const OpKernelInfo& info) {
   ORT_THROW_IF_ERROR(info.GetAttr("name", &name_));
   inplace_ = info.GetAttrOrDefault("inplace", static_cast<int64_t>(0));
   is_training_mode_ = static_cast<bool>(info.GetAttrOrDefault("training_mode", static_cast<int64_t>(0)));
-  ORT_THROW_IF_ERROR(info.GetAttr("call_convention", &call_convention_));
+  ORT_THROW_IF_ERROR(info.GetAttr("input_convention", &input_convention));
+  input_requires_grads_ = info.GetAttrsOrDefault("input_requires_grads", std::vector<int64_t>());
```
Comment (tlh20) [inline orttraining/orttraining/training_ops/cpu/torch/torch_custom_function_kernel_base.cc:34]: Should input_requires_grads always be defined?  If so we should omit the default and fail if it is missing (avoiding an extra allocation).

File: orttraining/orttraining/training_ops/cpu/torch/torch_custom_function_kernel_base.cc [hunk 12]
```diff
@@ -224,11 +224,14 @@ void PythonOpBase::SetOtherOutputs(OpKernelContext* context, std::vector<OrtValu
...
-  ORT_THROW_IF_ERROR(info.GetAttrs("input_tensor_types", input_tensor_types_));
+  ORT_THROW_IF_ERROR(info.GetAttr("output_convention", &output_convention_));
   ORT_THROW_IF_ERROR(info.GetAttrs("output_tensor_types", output_tensor_types_));
-  input_tensor_requires_grads_ = info.GetAttrsOrDefault("input_tensor_requires_grads", std::vector<int64_t>());
   output_tensor_requires_grads_ = info.GetAttrsOrDefault("output_tensor_requires_grads", std::vector<int64_t>());
-  SetPositions();
+  ORT_ENFORCE(output_tensor_types_.size() == output_tensor_requires_grads_.size(), "backward tensor output count mismatch");
+
+  // Exclude context and forward activations from inputs
+  const size_t tensor_input_count = (info.node().InputDefs().size() - 1) / 2;
```
Comment (tlh20) [inline orttraining/orttraining/training_ops/cpu/torch/torch_custom_function_kernel_base.cc:233]: This looks a bit fragile if we add extra inputs at some point in the future.  We could check that InputDefs().size() is odd as one safeguard -- are there other details we can assert?

**Review by tlh20 — CHANGES_REQUESTED** [review]
A few minor comments on places we could extend assertion checks.  Adjacent to some of the changes in this PR there is a use of "static" that looks suspicious -- could you check if that is correct?

File: orttraining/orttraining/core/graph/gradient_builder.cc [hunk 13]
```diff
@@ -1781,22 +1752,18 @@ IMPLEMENT_GRADIENT_BUILDER(GetPythonOpGradient) {
...
-    if (input_tensor_requires_grads.at(i)) {
+    if (src_attrs["output_tensor_requires_grads"].ints().Get(i - 1)) {
       // Only add FW outputs which
       //  1. are tensors,
       //  2. needs gradients (requires_grad=True in Pytorch).
       input_args.push_back(GO(i));
+    } else {
+      input_args.push_back(ArgDef());
     }
   }
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/gradient_builder.cc:1769]: Stashing all the outputs for backward computation could contribute to higher peak memory... 
I wonder if the seg fault is related to the ctx.save_for_backward(output_tensor)? 

This should be marked as an to-investigate issue.

File: orttraining/orttraining/training_ops/cpu/torch/torch_custom_function_kernel_base.cc (hunk 12 above)
Comment (SherlockNoMad) [reply orttraining/orttraining/training_ops/cpu/torch/torch_custom_function_kernel_base.cc:233]: +1. Please mark a todo here. 
Also, we are including all the outputs as a work around for the seg fault issue. 
After fix, the number of input should be just N+1 .

**Review by SherlockNoMad — DISMISSED** [review]

File: onnxruntime/core/language_interop_ops/torch/torch_proxy.cc (hunk 4 above)
Comment (pengwa) [reply onnxruntime/core/language_interop_ops/torch/torch_proxy.cc:180]: This is a good question. The dl_tensor is created by orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py (when we call to_dplack, which finally PyCapsule_New a PyObject*), so it should be only affected by our code there. User code should no increase its refcnt, but can increase the forward output torch tensor

File: orttraining/orttraining/core/graph/gradient_builder.cc (hunk 13 above)
Comment (pengwa) [reply orttraining/orttraining/core/graph/gradient_builder.cc:1769]: The seg fault happens when ctx is teared down before PythonOpGrad runs. To make sure ctx is not destroyed, at least one of those forward outputs remains before PythonOpGrad kernels completes. So you are right, we should not keep all outputs if we could identify which part of them contributes to backward computations. I will follow up on it.

Comment (pengwa) [reply orttraining/orttraining/core/graph/gradient_builder.cc:1769]: Added a TODO

File: orttraining/orttraining/core/graph/gradient_builder.cc (hunk 5 above)
Comment (pengwa) [reply orttraining/orttraining/core/graph/gradient_builder.cc:1777]: I understand the question this way (correct me if I am wrong): we need check the tensor + non-tensor inputs counts , for all attributes like "input_requires_grads , input_convention". For this, we have some checks when PythonOp is constructed. And I just added more checks enhance it.

File: orttraining/orttraining/core/graph/gradient_builder.cc [hunk 14]
```diff
@@ -1805,9 +1772,24 @@ IMPLEMENT_GRADIENT_BUILDER(GetPythonOpGradient) {
...
 
+  // src_attrs["input_requires_grads"] stores all inputs's requires_grad attributes,
+  // including both tensor inputs and non-tensor inputs (e.g. constants), here we filter out
+  // those non-tensor inputs when constructing PythonOpGrad's outputs.
+  const std::string& input_convention = src_attrs.at("input_convention").s();
+  const auto& fw_input_requires_grads = src_attrs["input_requires_grads"].ints();
+  std::vector<int64_t> bw_tensor_output_requires_grads;
+  for (auto i = 0; i < fw_input_requires_grads.size(); ++i) {
+    if (input_convention[i] == 'd') { // only handle gradients for tensor type inputs.
```
[the remark this answers is not in the document]
Comment (pengwa) [reply orttraining/orttraining/core/graph/gradient_builder.cc:1782]: Yes we did the check here. https://github.com/microsoft/onnxruntime/blob/7ed72bcdbff2033f913bba4126ffae082367d010/orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_exporter.py#L90 

If the value is not 'c' or 'd', we will fail it.

File: orttraining/orttraining/core/graph/training_op_defs.cc (hunk 6 above)
Comment (pengwa) [reply orttraining/orttraining/core/graph/training_op_defs.cc:2896]: Thanks alot for the formatting, my clang-format did not work at some point, let me format them in batch. :)

File: orttraining/orttraining/core/graph/training_op_defs.cc [hunk 15]
```diff
@@ -3086,34 +3089,25 @@ Return true if all elements are true and false otherwise.
...
-          AttributeProto::INTS,
-          false)
-      .Attr(
-          "input_tensor_requires_grads",
-          "Flags to indicate which inputs has gradient",
-          AttributeProto::INTS)
       .Attr(
           "output_tensor_types",
-          "Output types of autograd.Function.backward.",
+          "Output types of autograd.Function.backward outputs(including only tensor outputs).",
```
[the remark this answers is not in the document]
Comment (pengwa) [reply orttraining/orttraining/core/graph/training_op_defs.cc:3094]: my clang-format did not work at some point, I will check all format in batch. Thanks a lot!

File: orttraining/orttraining/training_ops/cpu/torch/torch_custom_function_kernel_base.cc (hunk 11 above)
Comment (pengwa) [reply orttraining/orttraining/training_ops/cpu/torch/torch_custom_function_kernel_base.cc:34]: Good suggestion! input_requires_grads  should always be defined, we should fail otherwise.

Comment (pengwa) [reply orttraining/orttraining/training_ops/cpu/torch/torch_custom_function_kernel_base.cc:34]: Fixed all

File: orttraining/orttraining/training_ops/cpu/torch/torch_custom_function_kernel_base.cc (hunk 12 above)
Comment (pengwa) [reply orttraining/orttraining/training_ops/cpu/torch/torch_custom_function_kernel_base.cc:233]: Thanks for the feedbacks, I would admit this is really not a good one.  Combining with Wei-Sheng's comment, I think a better way is, we add some redundant attributes like "input_tensor_types_" back to easily infer the tensor input count. That would be safe once we added more inputs. 

Besides, there are other pieces similar with this. I will try to eliminate this kind of logic once getting a better way to handle the lifecycle of ctx along with forward torch tensors.

Comment (pengwa) [reply orttraining/orttraining/training_ops/cpu/torch/torch_custom_function_kernel_base.cc:233]: Added a todo on my head

File: orttraining/orttraining/core/graph/training_op_defs.cc (hunk 2 above)
Comment (pengwa) [reply orttraining/orttraining/core/graph/training_op_defs.cc]: As we talked, I keep those attributes as redundancy, for more checks. Add some explanations in the op definitions in case of confusion of anyone curious why they are needed. :)

**Comment by pengwa** [comment]
> A few minor comments on places we could extend assertion checks. Adjacent to some of the changes in this PR there is a use of "static" that looks suspicious -- could you check if that is correct?

Thanks @tlh20  for your time reviewing , I have addressed all of comments . :)

**Review by tlh20 — APPROVED** [review]

**Review by baijumeswani — APPROVED** [review]
